### PR TITLE
West docs

### PR DESCRIPTION
**Pull request by mbolivar**
Add initial documentation for West.

This covers its design and scope, current and future plans, and API documentation for the runner library.

CC: @tejlmand

### CODE REVIEW

**Comment by mbolivar** [comment]
@galak better late than never, I hope :\

File: doc/west/flash-debug.rst [hunk 1]
```diff
@@ -0,0 +1,243 @@
+.. _west-flash-debug:
+
+West: Flashing and Debugging
+############################
+
+West provides three commands for running and interacting with Zephyr
+programs running on a board: ``flash``, ``debug``, and
+``debugserver``.
+
+These use information stored in the CMake cache to flash or attach a
```
Comment (ulfalizer) [inline doc/west/flash-debug.rst]: More of a comment, but as a CMake noob, I'd appreciate a short description of what the CMake cache is, or a link to the CMake documentation.

File: doc/west/index.rst [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+Additional features are under development for future versions of
+Zephyr:
+
+- Building Zephyr images.
+
+- Running Zephyr in emulation.
+
+- Bootloader integration: image building and signing, flashing
+  distinct images to different locations (the application is flashed
+  to separate area than the bootloader), and building and flashing the
```
Comment (ulfalizer) [inline doc/west/index.rst]: "a separate area"?

"the application and the bootloader are flashed to separate areas" maybe.

**Review by ulfalizer — CHANGES_REQUESTED** [review]

File: doc/conf.py [hunk 3]
```diff
@@ -15,9 +15,16 @@
 import sys
 import os
 
+if "ZEPHYR_BASE" not in os.environ:
+    sys.stderr.write("$ZEPHYR_BASE environment variable undefined.\n")
+    exit(1)
```
[the remark this answers is not in the document]
Comment (mbolivar) [reply doc/conf.py]: That's a bit unrelated to this patch series, no? Note that I'm just moving this code higher in the file. I Can add a cleanup patch on top if you'd like.

[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/conf.py]: Or just sneak it into the same patch I guess.

File: doc/west/index.rst (hunk 2 above)
Comment (mbolivar) [reply doc/west/index.rst]: IMO that would make the subsequent clause ("and building and flashing the bootloader itself") a bit redundant. But I agree that this is clumsily written. I'll try to streamline it.

Comment (mbolivar) [reply doc/west/index.rst]: Replaced with:

```
- Bootloader integration: bootloader-aware image building, signing,
  and flashing, as well as building and flashing the bootloader itself.
```

**Comment by mbolivar** [comment]
@ulfalizer thanks for review, comments addressed. I added a cleanup for the sys.exit shortcut, that's pretty neat!

**Review by carlescufi — APPROVED** [review]
Thanks Martí. This is great work. Very comprehensive and even covers the west design principles and constraints. I have no objections to the text as it stands.

**Review by ulfalizer — APPROVED** [review]

**Comment by mbolivar** [comment]
The true shippable failure is addressed in upstream west by https://github.com/zephyrproject-rtos/west/pull/4.

The other failure is spurious; there's no way I'm causing a QEMU riscv32 timing failure with this series.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**